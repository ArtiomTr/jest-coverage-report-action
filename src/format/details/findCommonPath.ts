import { dirname } from 'path';

/**
 * Find a common path from a list of filepaths.
 */
export const findCommonPath = (filepaths: string[]): string => {
    let commonRoot = '';
    if (filepaths.length) {
        // If the paths are sorted, any prefix common to all paths will be common to the sorted first and last strings.
        const sortedPaths = [...filepaths].sort();
        const first = sortedPaths[0];
        const last = sortedPaths[sortedPaths.length - 1];

        const len = Math.min(first.length, last.length);
        for (let i = 0; i < len; i++) {
            const ch1 = first[i];
            const ch2 = last[i];
            if (ch1 == ch2) {
                commonRoot += ch1;
            } else {
                break;
            }
        }

        // If it doesn't appear to be a directory (partial filename), get the parent directory
        if (commonRoot.length && commonRoot[commonRoot.length - 1] !== '/') {
            commonRoot = dirname(commonRoot) + '/';
        }
    }
    return commonRoot;
};
