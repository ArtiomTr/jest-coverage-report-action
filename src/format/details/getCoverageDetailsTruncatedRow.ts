import { i18n } from '../../utils/i18n';

export const getCoverageDetailsTruncatedRow = (
    rowCount: number
): Array<string> => {
    return [
        '',
        i18n('detailsTruncatedRows', { truncatedRows: rowCount }),
        '',
        '',
        '',
        '',
    ];
};
