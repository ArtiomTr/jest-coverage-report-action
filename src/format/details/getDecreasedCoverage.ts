import { CoverageDetail, CoverageDetailsMap } from '../../typings/Coverage';

const coverageLessThan = (first: CoverageDetail, second: CoverageDetail) =>
    first.statements < second.statements ||
    first.branches < second.branches ||
    first.functions < second.functions;

export const getDecreasedCoverage = (
    headDetails: CoverageDetailsMap,
    baseDetails: CoverageDetailsMap | undefined
) =>
    Object.keys(headDetails)
        .filter(
            (filename) =>
                headDetails[filename] &&
                baseDetails?.[filename] &&
                coverageLessThan(headDetails[filename], baseDetails[filename])
        )
        .reduce<{
            headDetails: CoverageDetailsMap;
            baseDetails: CoverageDetailsMap;
        }>(
            (acc, filename) => {
                acc.headDetails[filename] = headDetails[filename];
                acc.baseDetails[filename] = baseDetails![filename];
                return acc;
            },
            { headDetails: {}, baseDetails: {} }
        );
