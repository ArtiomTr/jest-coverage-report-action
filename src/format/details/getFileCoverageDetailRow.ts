import { shrinkLongPath } from './shrinkLongPath';
import { CoverageDetail } from '../../typings/Coverage';
import { formatPercentage } from '../../utils/formatPercentage';
import { getStatusOfPercents } from '../../utils/getStatusOfPercents';

export const getFileCoverageDetailRow = (
    filename: string,
    headDetail: CoverageDetail,
    baseDetail?: CoverageDetail,
    threshold?: number
): Array<string> => [
    getStatusOfPercents(headDetail.lines, threshold),
    shrinkLongPath(filename),
    formatPercentage(headDetail.statements, baseDetail?.statements),
    formatPercentage(headDetail.branches, baseDetail?.branches),
    formatPercentage(headDetail.functions, baseDetail?.functions),
    formatPercentage(headDetail.lines, baseDetail?.lines),
];
