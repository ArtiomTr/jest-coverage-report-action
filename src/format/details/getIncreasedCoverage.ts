import { CoverageDetail, CoverageDetailsMap } from '../../typings/Coverage';

const coverageGreaterThan = (first: CoverageDetail, second: CoverageDetail) =>
    first.statements > second.statements ||
    first.branches > second.branches ||
    first.functions > second.functions;

export const getIncreasedCoverage = (
    headDetails: CoverageDetailsMap,
    baseDetails: CoverageDetailsMap | undefined
) =>
    Object.keys(headDetails)
        .filter(
            (filename) =>
                headDetails[filename] &&
                baseDetails?.[filename] &&
                coverageGreaterThan(headDetails[filename], baseDetails[filename])
        )
        .reduce<{
            headDetails: CoverageDetailsMap;
            baseDetails: CoverageDetailsMap;
        }>(
            (acc, filename) => {
                acc.headDetails[filename] = headDetails[filename];
                acc.baseDetails[filename] = baseDetails![filename];
                return acc;
            },
            { headDetails: {}, baseDetails: {} }
        );
