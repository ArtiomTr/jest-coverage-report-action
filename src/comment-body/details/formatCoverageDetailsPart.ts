import markdownTable from 'markdown-table';

import { getFileCoverageDetailRow } from './getFileCoverageDetailRow';
import { ParsedCoverageDetails } from '../../collect-coverage/parseCoverageDetails';
import { details } from '../strings.json';
import { createMarkdownSpoiler } from '../utils/createMarkdownSpoiler';
import { formatHeadingAndTable } from '../utils/formatHeadingAndTable';

export type DetailsFormatOptions = {
    summary: string;
    heading: string;
};

export const formatCoverageDetailsPart = (
    formatOptions: DetailsFormatOptions,
    headDetails: ParsedCoverageDetails,
    baseDetails?: ParsedCoverageDetails
): string | undefined => {
    const { summary, heading } = formatOptions;

    const tableContent = Object.keys(headDetails).map((filename) =>
        getFileCoverageDetailRow(
            filename,
            headDetails[filename],
            baseDetails?.[filename]
        )
    );

    if (tableContent.length > 0) {
        return createMarkdownSpoiler({
            body: formatHeadingAndTable(
                heading,
                markdownTable([details.columnHeaders, ...tableContent])
            ),
            summary,
        });
    }

    return undefined;
};
