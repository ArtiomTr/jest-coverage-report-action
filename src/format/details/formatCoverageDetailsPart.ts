import markdownTable from 'markdown-table';

import { getFileCoverageDetailRow } from './getFileCoverageDetailRow';
import { CoverageDetailsMap } from '../../typings/Coverage';
import { createMarkdownSpoiler } from '../../utils/createMarkdownSpoiler';
import { i18n } from '../../utils/i18n';
import { withExplanation } from '../../utils/withExplanation';

export const formatCoverageDetailsPart = (
    summary: string,
    headDetails: CoverageDetailsMap,
    baseDetails?: CoverageDetailsMap,
    threshold?: number
): string | undefined => {
    const tableContent = Object.keys(headDetails).map((filename) =>
        getFileCoverageDetailRow(
            filename,
            headDetails[filename],
            baseDetails?.[filename],
            threshold
        )
    );

    if (tableContent.length > 0) {
        return createMarkdownSpoiler({
            body: markdownTable(
                [
                    [
                        withExplanation(
                            i18n('status'),
                            i18n('statusExplanation')
                        ),
                        i18n('filename'),
                        i18n('statements'),
                        i18n('branches'),
                        i18n('functions'),
                        i18n('lines'),
                    ],
                    ...tableContent,
                ],
                {
                    align: ['c', 'l', 'l', 'l', 'l', 'l'],
                }
            ),
            summary,
        });
    }

    return undefined;
};
