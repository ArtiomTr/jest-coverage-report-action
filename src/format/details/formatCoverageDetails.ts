import { formatCoverageDetailsPart } from './formatCoverageDetailsPart';
import { getDecreasedCoverage } from './getDecreasedCoverage';
import { getNewFilesCoverage } from './getNewFilesCoverage';
import { CoverageDetailsMap } from '../../typings/Coverage';
import { i18n } from '../../utils/i18n';

export const formatCoverageDetails = (
    headDetails: CoverageDetailsMap,
    baseDetails: CoverageDetailsMap | undefined,
    threshold: number | undefined
): string => {
    const decreasedCoverage = getDecreasedCoverage(headDetails, baseDetails);

    return [
        formatCoverageDetailsPart(
            i18n('newFilesCoverage'),
            getNewFilesCoverage(headDetails, baseDetails),
            undefined,
            threshold
        ),
        formatCoverageDetailsPart(
            i18n('decreasedCoverageFiles'),
            decreasedCoverage.headDetails,
            decreasedCoverage.baseDetails,
            threshold
        ),
    ].join('\n');
};
