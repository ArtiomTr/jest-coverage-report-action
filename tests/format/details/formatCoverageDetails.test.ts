import { formatCoverageDetails } from '../../../src/format/details/formatCoverageDetails';

describe('formatCoverageDetails', () => {
    it('should match snapshots', () => {
        expect(
            formatCoverageDetails(
                {
                    'increased.ts': {
                        lines: 80,
                        statements: 70,
                        branches: 50,
                        functions: 50,
                        filename: 'increased.ts',
                    },
                    'decreased.ts': {
                        lines: 80,
                        statements: 70,
                        branches: 50,
                        functions: 50,
                        filename: 'decreased.ts',
                    },
                    'newFile.ts': {
                        lines: 50,
                        statements: 50,
                        branches: 50,
                        functions: 50,
                        filename: 'newFile.ts',
                    },
                },
                {
                    'increased.ts': {
                        lines: 50,
                        statements: 50,
                        branches: 30,
                        functions: 30,
                        filename: 'increased.ts',
                    },
                    'decreased.ts': {
                        lines: 90,
                        statements: 80,
                        branches: 60,
                        functions: 60,
                        filename: 'decreased.ts',
                    },
                },
                70
            )
        ).toMatchSnapshot();
    });
});
