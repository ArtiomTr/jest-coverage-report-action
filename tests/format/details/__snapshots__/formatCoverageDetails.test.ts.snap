// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatCoverageDetails > should match snapshots 1`] = `
"
<details><summary>Show new covered files 游냒</summary>
<br/>

| <div title=\\"Status of coverage:&#10; 游릭 - ok&#10; 游리 - slightly more than threshold&#10; 游댮 - under the threshold\\">St.<sup>:grey_question:</sup></div> | File       | Statements | Branches | Functions | Lines |
| :----------------------------------------------------------------------------------------------------------------------------------------------------: | :--------- | :--------- | :------- | :-------- | :---- |
|                                                                           游댮                                                                           | newFile.ts | 50%        | 50%      | 50%       | 50%   |

</details>


<details><summary>Show files with reduced coverage 游댵</summary>
<br/>

| <div title=\\"Status of coverage:&#10; 游릭 - ok&#10; 游리 - slightly more than threshold&#10; 游댮 - under the threshold\\">St.<sup>:grey_question:</sup></div> | File         | Statements                                             | Branches                                               | Functions                                              | Lines                                                  |
| :----------------------------------------------------------------------------------------------------------------------------------------------------: | :----------- | :----------------------------------------------------- | :----------------------------------------------------- | :----------------------------------------------------- | :----------------------------------------------------- |
|                                                                           游리                                                                           | decreased.ts | <div title=\\"Base coverage is: 80%\\">70% (-10% 游댵)</div> | <div title=\\"Base coverage is: 60%\\">50% (-10% 游댵)</div> | <div title=\\"Base coverage is: 60%\\">50% (-10% 游댵)</div> | <div title=\\"Base coverage is: 90%\\">80% (-10% 游댵)</div> |

</details>
"
`;
